plugin/pkg/scheduler/core/equivalence_cache.go

```diff
@@ -61,74 +66,155 @@ func NewEquivalenceCache(getEquivalencePodFunc algorithm.GetEquivalencePodFunc)
+	if v, ok := ec.algorithmCache[nodeName].predicatesCache.Get(predicateKey); ok {
+		predicateMap := v.(PredicateMap)
+		predicateMap[equivalenceHash] = predicateItem
+		ec.algorithmCache[nodeName].predicatesCache.Add(predicateKey, predicateMap)
+
+	} else {
+		ec.algorithmCache[nodeName].predicatesCache.Add(predicateKey,
+			PredicateMap{
+				equivalenceHash: predicateItem,
+			})
 	}
+	glog.V(5).Infof("Updated cached predicate: %v on node: %s, with item %v", predicateKey, nodeName, predicateItem)
 }
 
-// GetCachedPredicates gets cached predicates for equivalence class
-func (ec *EquivalenceCache) GetCachedPredicates(pod *v1.Pod, nodes []*v1.Node) ([]*v1.Node, FailedPredicateMap, []*v1.Node) {
-	fitNodeList := []*v1.Node{}
-	failedPredicates := FailedPredicateMap{}
-	noCacheNodeList := []*v1.Node{}
-	equivalenceHash := ec.hashEquivalencePod(pod)
-	for _, node := range nodes {
-		findCache := false
-		if algorithmCache, exist := ec.algorithmCache[node.Name]; exist {
-			if cachePredicate, exist := algorithmCache.predicatesCache.Get(equivalenceHash); exist {
-				hostPredicate := cachePredicate.(HostPredicate)
+// TryPredicateWithECache returns:
+// 1. if fit
+// 2. reasons if not fit
+// 3. if this cache is invalid
+// based on cached predicate results
+func (ec *EquivalenceCache) PredicateWithECache(pod *v1.Pod, nodeName, predicateKey string, equivalenceHash uint64) (bool, []algorithm.PredicateFailureReason, bool) {
+	ec.RLock()
+	defer ec.RUnlock()
+	if algorithmCache, exist := ec.algorithmCache[nodeName]; exist {
+		if cachePredicate, exist := algorithmCache.predicatesCache.Get(predicateKey); exist {
+			predicateMap := cachePredicate.(PredicateMap)
+			// TODO(harry) In this case, the first pod will use more time
+			// because it does not have ecache item but has to run into here,
+			// but in this case invalidate functions are more efficient.
+			// TODO(harry) Is it possible for cache is not updated immediately?
+			if hostPredicate, ok := predicateMap[equivalenceHash]; ok {
+				if hostPredicate.IsInvalid {
+					return false, []algorithm.PredicateFailureReason{}, true
+				}
 				if hostPredicate.Fit {
-					fitNodeList = append(fitNodeList, node)
+					return true, []algorithm.PredicateFailureReason{}, false
 				} else {
-					failedPredicates[node.Name] = hostPredicate.FailReasons
+					return false, hostPredicate.FailReasons, false
 				}
-				findCache = true
 			}
+			glog.V(5).Infof("Calculated predicate: %v for pod: %s on node: %s based on equivalence cache", predicateKey, pod.GetName, nodeName)
 		}
-		if !findCache {
-			noCacheNodeList = append(noCacheNodeList, node)
+	}
+	return false, []algorithm.PredicateFailureReason{}, true
+}
+
+// InvalidCachedPredicateItem marks all items of given predicateKeys, of all pods, on the given node as invalid
+func (ec *EquivalenceCache) InvalidCachedPredicateItem(nodeName string, predicateKeys sets.String) {
```

Comment:
I would change the name to "InvalidatePredicatesOfNode". "Invalid" is confusing and gives the impression that you are just checking a condition.